clang/include/clang/Lex/Preprocessor.h

```diff
@@ -2835,6 +2835,9 @@ class Preprocessor {
     if (Identifier.getIdentifierInfo()->isRestrictExpansion() &&
         !SourceMgr.isInMainFile(Identifier.getLocation()))
       emitRestrictExpansionWarning(Identifier);
+
+    if (Identifier.getIdentifierInfo()->getName() == "INFINITY")
```

Comment:
What about `NAN`?

Reply:
Because there are other C standard libraries which do something different. For example, musl: https://git.musl-libc.org/cgit/musl/tree/include/math.h#n18

Reply:
> Since NAN is defined as (-(float)(INFINITY * 0.0F)) when used in the program it will trigger the "use of infinity via a macro results in undefined behavior due to the currently enabled floating-point options [-Wnan-and-infinity-disabled]". Wouldn't that be enough? If I add NaN here we would get 2 warnings.

This is actually a bit unfortunate. If I use "-fhonor-nans -fno-honor-infinities" this definition of NAN will trigger a warning, even though I expect that we'll constant-fold it as expected without regard to the "no-honor-infinities" setting.

Reply:
We will?
```
#ifndef _HUGE_ENUF
    #define _HUGE_ENUF  1e+300  // _HUGE_ENUF*_HUGE_ENUF must overflow
#endif

#define INFINITY   ((float)(_HUGE_ENUF * _HUGE_ENUF))
...
#define NAN        ((float)(INFINITY * 0.0F))
```
the multiplication for `INFINITY` won't form an actual infinity value in that case, so why would the multiplication with `0.0` not generate a NAN macro that folds to 0.0F?

If use of `INFINITY` is undefined behavior when `-fno-honor-infinities` is passed... it makes sense to me that the definition of `NAN` using `INFINITY` is UB by the same logic.